// satteri@0.9.4/dist/hast/generated/node-types.js
// @generated by `cargo run -p satteri-layout-codegen`. Do not edit by hand.
/** Node-type tag -> canonical AST name. */
export const TYPE_NAMES = {
    0: "root",
    1: "element",
    2: "text",
    3: "comment",
    4: "doctype",
    5: "raw",
    10: "mdxJsxFlowElement",
    11: "mdxJsxTextElement",
    12: "mdxFlowExpression",
    13: "mdxjsEsm",
    14: "mdxTextExpression",
};
/** Canonical AST name -> node-type tag. */
export const NAME_TO_TYPE = {
    root: 0,
    element: 1,
    text: 2,
    comment: 3,
    doctype: 4,
    raw: 5,
    mdxJsxFlowElement: 10,
    mdxJsxTextElement: 11,
    mdxFlowExpression: 12,
    mdxjsEsm: 13,
    mdxTextExpression: 14,
};
/** Names a plugin can subscribe to (every node except `root`). */
export const VISITOR_KEYS = new Set([
    "element",
    "text",
    "comment",
    "doctype",
    "raw",
    "mdxJsxFlowElement",
    "mdxJsxTextElement",
    "mdxFlowExpression",
    "mdxjsEsm",
    "mdxTextExpression",
]);
/** Name -> tag for the types the op-stream can encode; one lookup gates AND
 *  resolves the emit-path tag. Excluded names (see `*_OPSTREAM_EXCLUDED` in
 *  schema.rs) have no encoding — the visitor throws for them. */
export const HAST_OPSTREAM_TYPES = {
    element: 1,
    text: 2,
    comment: 3,
    raw: 5,
    mdxJsxFlowElement: 10,
    mdxJsxTextElement: 11,
    mdxFlowExpression: 12,
    mdxjsEsm: 13,
    mdxTextExpression: 14,
};
/** Names of the variable-length `custom` node types (hand-written or
 *  `Tail`-generated codec). The op-stream round-trip oracle asserts it covers
 *  every one, so a forgotten or drifted encode/decode arm fails loudly. */
export const HAST_CUSTOM_TYPES = [
    "element",
    "mdxJsxFlowElement",
    "mdxJsxTextElement",
];
